google/cloud/bigquery/job/query.py

```diff
@@ -989,19 +989,30 @@ def done(self, retry=DEFAULT_RETRY, timeout=None, reload=True):
 
         Returns:
             bool: True if the job is complete, False otherwise.
+
+        Raises:
+            google.api_core.exceptions,GoogleAPICallError:
+                If a non-retryable error ocurrs while reloading the job metadata,
+                or if too many retry attempts fail.
```

Comment:
Despite catching and handling reload-query-results errors, we should nevertheless let any job reload errors bubble up, IMO. We don't really know how to handle them and what to return if we cannot even get job info in the first place.

Reply:
Hmmm... This is a departure from https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future.done which doesn't throw any exceptions.

I agree that bubbling up the error would give more information to the user that we otherwise wouldn't know how to interpret, though. The alternative would be to return `False`, but then I think we risk infinite loops by our blocking poll that uses this method.

Or can we call `set_exception` and return False in this case? We'd want a unit test that confirms `result()` doesn't get into an infinite loop.

Reply:
As discussed in the chat, returning `False` would cause the `done()` method to be repeated when run as part of a `_blocking_poll()`, even if exception was set on the job. However, we can still set the exception and return `True` to terminate the polling.

Terminating the polling is actually fine, because `self._reload_query_results()` and `self.reload()` already do their own retry cycle and if that has failed, we should give up and signal that the job instance - which is actually _future_ - is done, even if the result of the future is an exception.

We must also catch any errors that happen on job reload itself and set the job/future state accordingly.